### PR TITLE
KTOR-4770 Fix ignoreBody for Server ContentNegotiation Request

### PR DESCRIPTION
**Pull request by e5l**
Fix [KTOR-4770](https://youtrack.jetbrains.com/issue/KTOR-4770)

### CODE REVIEW

File: ktor-shared/ktor-serialization/common/src/ContentConverter.kt [hunk 1]
```diff
@@ -119,8 +119,7 @@ public suspend fun List<ContentConverter>.deserialize(
     // 1. there is no suitable converter
     // 2. result of deserialization is null
     // We can differentiate these cases by checking if body was consumed or not
-    val result = this.asFlow()
-        .filter { !body.isClosedForRead }
```
Comment (rsinukov) [inline ktor-shared/ktor-serialization/common/src/ContentConverter.kt:123]: This check is needed when some converter has consumed body and returned `null`. The next converter will crash with EOF if it tries to parse.

Comment (rsinukov) [reply ktor-shared/ktor-serialization/common/src/ContentConverter.kt:123]: you can move it to `firstOrNull { it != null || body.isClosedForRead }`. Now it's confusing.

File: ktor-server/ktor-server-test-host/jvmAndNix/src/io/ktor/server/testing/client/TestHttpClientEngine.kt [hunk 2]
```diff
@@ -93,27 +93,14 @@ public class TestHttpClientEngine(override val config: TestHttpClientConfig) : H
...
-        headers.flattenForEach { name, value ->
-            if (HttpHeaders.ContentLength == name) return@flattenForEach // set later
-            if (HttpHeaders.ContentType == name) return@flattenForEach // set later
-            addHeader(name, value)
-        }
-
-        content.headers.flattenForEach { name, value ->
-            if (HttpHeaders.ContentLength == name) return@flattenForEach // TODO: throw exception for unsafe header?
-            if (HttpHeaders.ContentType == name) return@flattenForEach
+        mergeHeaders(headers, content) { name, value ->
```
Comment (rsinukov) [inline ktor-server/ktor-server-test-host/jvmAndNix/src/io/ktor/server/testing/client/TestHttpClientEngine.kt:101]: Is it related or just a cleanup?

**Review by rsinukov — APPROVED** [review]

File: ktor-server/ktor-server-test-host/jvmAndNix/src/io/ktor/server/testing/client/TestHttpClientEngine.kt (hunk 2 above)
Comment (e5l) [reply ktor-server/ktor-server-test-host/jvmAndNix/src/io/ktor/server/testing/client/TestHttpClientEngine.kt:101]: it's related

File: ktor-shared/ktor-serialization/common/src/ContentConverter.kt (hunk 1 above)
Comment (e5l) [reply ktor-shared/ktor-serialization/common/src/ContentConverter.kt:123]: sure, let me check

File: ktor-shared/ktor-serialization/common/src/ContentConverter.kt [hunk 3]
```diff
@@ -119,19 +119,18 @@ public suspend fun List<ContentConverter>.deserialize(
...
-        .map { converter ->
-            converter.deserialize(
-                charset = charset,
-                typeInfo = typeInfo,
-                content = body
-            )
-        }
-        .firstOrNull { it != null }
+    val result = asFlow().map { converter ->
```
Comment (rsinukov) [inline ktor-shared/ktor-serialization/common/src/ContentConverter.kt:122]: personally, i prefer formatting when each operator is on a separate line
```
asFlow()
  .map { 
      ...
  }
  .filter { ... }
```

**Pull request merged**